### PR TITLE
First assembler instructions

### PR DESCRIPTION
**Pull request by AntonKozlov**
A lot of work done for encodings, great!

Can we merge this?

### CODE REVIEW

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -209,7 +209,51 @@ class Assembler : public AbstractAssembler {
...
+  static inline int rs3(int x) { return x << 27; }
+  static inline int rs3(Register x) { return rs3(x->encoding()); }
+
+  // TODO_RISCV: more asserts for immX() functions?
+  static inline int immi(int x) { return (int)((unsigned)x << 20); }
+  static inline int imms(int x) { return (int)(((x & 0x1f) << 7) | ((unsigned)x >> 5 << 25)); }
+  static inline int immb(int x) {
+    assert((x & 1) == 0, "B-immediate must be zero in bit 0");
+    return (int)(
+        ((x & 0x1e) << 7) | ((x & 0x7e0) >> 5 << 25) |
```
Comment (Geckostya) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp]: I think that `((x >> 5) & 0x3f) << 25` is more readable when `((x & 0x7e0) >> 5 << 25)`

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 2]
```diff
@@ -209,7 +209,51 @@ class Assembler : public AbstractAssembler {
...
+        ((unsigned)(x >> 12) << 31) | (((x >> 11) & 0x1) << 6)
```
Comment (Geckostya) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp]: `(((x >> 11) & 0x1) << 6)` it should be in 7th bit, not 6th

**Review by Geckostya — CHANGES_REQUESTED** [review]
Add more RISCV specification links, it is hard to find code to docs accordance

**Comment by DmKazakov** [comment]
I have a question about naming convention. For example, how should method corresponding to `FCVT.D.Q` be named? `fcvtdq`? `fcvt_d_q`? Something else?

Also, I need some verification regarding floating point instructions. Some of them have `rm` parameter that defines rounding mode. All I can find in the manual about it:
>All floating-point to integer and integer to floating-point conversion instructions round according
to the rm field.

So it's not clear to me whether `rm` should be a register or an immediate value.

**Comment by AntonKozlov** [comment]
> I have a question about naming convention. For example, how should method corresponding to `FCVT.D.Q` be named? `fcvtdq`? `fcvt_d_q`? Something else?

I have no strong preference. `fcvtd_dq` is also possible

> Also, I need some verification regarding floating point instructions. Some of them have `rm` parameter that defines rounding mode. All I can find in the manual about it:
> 
> > All floating-point to integer and integer to floating-point conversion instructions round according
> > to the rm field.
> 
> So it's not clear to me whether `rm` should be a register or an immediate value.

A far as I can see `rm` occupies bits 12-14, so it should be 3-bit immediate value

**Review by Geckostya — APPROVED** [review]

**Pull request merged**